### ISSUE TITLE
fix(nargs): Rust counts the self receiver as a parameter

### ISSUE THREAD

**Issue by dekobon** [dekobon/big-code-analysis#457]
## Summary

`RustCode::is_non_arg` does not exclude `Rust::SelfParameter`, so the default `compute_args` counts `self` / `&self` / `&mut self` / `self: Box<Self>` as a formal parameter. Go and C++ correctly exclude their receivers, so Rust nargs diverges and over-counts every method.

## Location

- `src/checker.rs:937-942` — `RustCode::is_non_arg` matches `LPAREN|COMMA|RPAREN|PIPE|AttributeItem`, omits `SelfParameter`
- `src/metrics/nargs.rs` (default `compute_args`) — counts `parameters` children not flagged `is_non_arg`

## Evidence

```rust
struct S; impl S {
  fn a(self) {}
  fn b(&mut self, x: i32) {}
  fn c(self: Box<Self>, x: i32, y: i32) {}
}
```
- Correct: a=0, b=1, c=2 → `fn_nargs_sum=3`, `fn_nargs_max=2`.
- Measured: `fn_nargs_sum=6`, `fn_nargs_max=3` — each receiver counted.

Untested: every Rust nargs test uses free functions or closures; `tests/nargs_cross_language_parity.rs` uses a receiver-less `fn f(a,b,c)`. Diverges from Go (`go_method_excludes_receiver`) and C++ (`cpp_member_function_this_not_counted`).

## Expected Behavior

The `self` receiver is not a parameter (matching Go/C++).

## Actual Behavior

`self`/`&self`/`&mut self`/`self: T` each add 1 to nargs.

## Impact

`nargs_sum`, `nargs_max`, and `nargs_average` (which divides by the function count) inflated for every Rust method; breaks the documented cross-language nargs contract.

---

## Resolution

Fixed on `fix/issue-457` (commit b218a153). `RustCode::is_non_arg`
now excludes `Rust::SelfParameter`, so `self` / `&self` / `&mut self`
are no longer counted — matching Go and C++ receiver handling.

Correction to the fix plan: `self: Box<Self>` parses as an ordinary
`parameter` (it carries a type annotation), not `self_parameter`, so
method `c` reports 3 args, not 2. Final counts: `a=0, b=1, c=3, d=0`.
Counting the typed receiver matches the grammar's own modeling.

Test `rust_method_excludes_self_receiver` covers all three bare forms
(revert-verified against the 6/3 over-count) and pins the typed-receiver
case. Corpus snapshots (submodule commit 8124ec66) and the self-scan
baseline were refreshed.

Separate bug found (filed independently): Java's explicit receiver
`void m(S this, int a)` is also miscounted (`ReceiverParameter` missing
from `JavaCode::is_non_arg`).

**Comment by dekobon** [comment]
## Fix Plan

### Root Cause
`is_non_arg` omits `SelfParameter`.

### Implementation Steps
1. Add `Rust::SelfParameter` to `RustCode::is_non_arg`.

### Tests to Add
- `rust_method_excludes_self_receiver` — the three-method example → sum=3, max=2 (revert-verified against 6/3); cover `self`, `&self`, `&mut self`, `self: Box<Self>`.

### Verification
- [ ] `cargo test --workspace --all-features` clean
- [ ] Submodule snapshots accepted/pushed + pointer bump (Rust nargs shift)
- [ ] Rust method nargs matches Go/C++ parity fixtures

**Comment by dekobon** [comment]
Fixed in `fix/issue-457` (commit b218a153).

## What changed
`RustCode::is_non_arg` now excludes the `self_parameter` node, so
`self`, `&self`, and `&mut self` no longer count as formal parameters
— mirroring Go's `receiver`-field exclusion and C++'s implicit-`this`
exclusion. The fix is the one-token addition from the plan, plus a
clarifying comment.

## Correction to the fix plan
The example's method `c(self: Box<Self>, x, y)` does **not** drop to
2 args. A typed receiver written `self: Box<Self>` parses as an
ordinary `parameter` node (it carries a type annotation), **not** as
`self_parameter`:

```
fn c(self: Box<Self>, …)   ->  (parameter "self: Box<Self>") (parameter "x") (parameter "y")
fn b(&mut self, …)         ->  (self_parameter "&mut self")   (parameter "x")
```

So after the fix the per-method counts are:

| method | form | nargs (before) | nargs (after) |
|--------|------|---------------:|--------------:|
| `a(self)` | `self` | 1 | **0** |
| `b(&mut self, x)` | `&mut self` | 2 | **1** |
| `c(self: Box<Self>, x, y)` | typed receiver | 3 | **3** (unchanged) |
| `d(&self)` | `&self` | 1 | **0** |

Excluding `self_parameter` only is the correct, grammar-faithful
scope; counting the typed `self: Box<Self>` receiver matches the
tree-sitter grammar's own modeling and upstream rust-code-analysis.

## Test
`rust_method_excludes_self_receiver` (in `src/metrics/nargs.rs`)
covers all three bare forms (`self`, `&self`, `&mut self`) →
`fn_args_sum == 3`, `fn_args_max == 2`, and pins the typed-receiver
case separately. Revert-verified: against pre-fix code the test fails
with `sum = 6` (the over-count from the issue).

## Snapshots / baseline
Corpus snapshots (submodule `dekobon/big-code-analysis-output`,
commit 8124ec66) and the `csv_rust_counter` snapshot were regenerated
— metric-value-only nargs shifts for Rust fixtures with methods. The
self-scan baseline was refreshed.

## Out of scope (separate bug found)
Java's **explicit** receiver parameter (`void m(S this, int a)`) is
also miscounted — `JavaCode::is_non_arg` omits `ReceiverParameter`,
so that method reports `nargs = 2` instead of 1. Kotlin has a
`Receiver` kind worth checking too. Filing separately to keep this
commit atomic to Rust.

**Comment by dekobon** [comment]
Wave 6 QA follow-up: the typed-receiver gap is now closed.

The initial fix excluded only the bare `self_parameter` node
(`self`, `&self`, `&mut self`). A **typed** receiver
(`self: Box<Self>`, `self: Rc<Self>`, `self: Pin<&mut Self>`) parses
as an ordinary `parameter` node, so it was still counted — deviating
from this issue's stated acceptance (`fn c(self: Box<Self>, x, y)`
reported 3, should be 2).

The QA pass added structural detection: a `Rust::Parameter` whose
binding is the `self` keyword (`Rust::Zelf`, kind 142) is now also
excluded. This is unambiguous because the `self` keyword node (142) is
distinct from the `Self` type-identifier (354), verified by AST dump:

- `self: Pin<&mut Self>` -> has a direct `{self:142}` child -> excluded.
- `Self { x }: Self` (struct-pattern param) -> `type_identifier:354`,
  no `self:142` -> correctly counted.
- `s: S` / `freefn(s, t)` -> no `self` child -> correctly counted.

`fn c(self: Box<Self>, x, y)` now reports `nargs = 2`, matching the
acceptance criteria and Go/C++ receiver parity. Regression test
`rust_method_excludes_self_receiver` covers `self`/`&self`/`&mut self`
plus all three typed forms; it revert-fails against pre-fix code.

No corpus snapshot shift (the integration corpus contains no typed-self
receivers).

While verifying receiver parity across languages, found that Java's
explicit receiver parameter (`void m(S this, int a)`) is over-counted
the same way — filed separately as #470. Kotlin was checked and is not
affected (extension receiver sits outside the parameter list).
